File: python/algorithms/TwoSum/src/two_sum.py
""" Two Sum Solution. """

from typing import List


class TwoSum:
    def run_two_sum(self, nums: List[int], target: int) -> List[int]:
        previous_nums = {}

        for idx, ele in enumerate(nums):

            # Find compliment number
            compliment_num = target - ele

            # If the compliment is already in the values, return indices
            if (
                compliment_num in previous_nums.keys()
                and idx not in previous_nums.values()
            ):
                return [previous_nums[compliment_num], idx]
            else:
                previous_nums[ele] = idx

        else:
            print(f"No two numbers found to add to {target}")


"""
Explanation:

- Iterating through a list costs O(n) time complexity.

- Using a brute force method would cost O(n^2) time complexity.

- Using a dictionary (HashMap) to store the previous numbers we come across, can reduce the time complexity.

--- Why does it reduce the time complexity? ---
    --> Because to look up a value in a dictionary take O(1) time complexity
    --> So looking up the compliment number in the dictionary for each element in the list costs 0(n*1) = O(n) time complexity since we drop constants.
"""

File: java/algorithms/TwoSum/src/main/java/org/github/Thorpe7/TwoSum.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 */
package org.github.Thorpe7;

/**
 *
 * @author Thorpe7
 */
import java.util.HashMap;
import java.util.Map;

class TwoSum {

    public int[] runTwoSum(int[] nums, int target) {

        // Start by creating an empty hashmap
        Map<Integer, Integer> usedNumbers = new HashMap<>();

        // Iterate through the array
        for (int i = 0; i < nums.length; i++) {
            int complementNum = target - nums[i];

            // If the complement is in the map, we found the answer
            if (usedNumbers.containsKey(complementNum)) {
                // Define the answer array
                int[] twoSumSolutionIdxs = {i, usedNumbers.get(complementNum)};
                return twoSumSolutionIdxs;
            } else {
                usedNumbers.put(nums[i], i);
            }
        }
        // If no solution is found, return type in function signature is int[]
        return new int[0];
    }

    public static void main(String[] args) {
    }
}

/**
 * Explanation:
 *
 *  *- Iterating through a list costs O(n) time complexity.
 *
 *  *- Using a brute force method would cost O(n^2) time complexity.
 *
 *  *- Using a dictionary (HashMap) to store the previous numbers we come
 * across, can reduce the time complexity.
 *
 *  *--- Why does it reduce the time complexity? ---
 *
 * Because to look up a value in a dictionary takes O(1) time complexity. So
 * looking up the compliment number in the dictionary for each element in the
 * list costs 0(n*1) = O(n) time complexity since we drop constants.
 */
